text/0436-gamelift-l2.md

```diff
@@ -0,0 +1,945 @@
+
+This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project. It allows you to define components for your matchmaking configuration or gmae server fleet management system.
+
+#### GameLift FlexMatch
+
+##### Defining a Matchmaking configuration
+
+FlexMatch is available both as a GameLift game hosting solution (including Realtime Servers) and as a standalone matchmaking service. To set up a FlexMatch matchmaker to process matchmaking requests, you have to create a matchmaking configuration based on a RuleSet.
+
+More details about matchmaking ruleSet are covered [below](#ruleSet).
+
+There is two types of Matchmaking configuration: through a queue system to let FlexMatch forms matches and uses the specified GameLift queue to start a game session for the match, and through a standalone version to let FlexMatch forms matches and returns match information in an event.
+
+
+Either a Standalone Matchmaking configuration
+
+```ts
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+
+new gamelift.MatchmakingConfiguration(this, 'Standalone Matchmaking', {
+  requestTimeouts: Duration.seconds(35)
+});
+```
+
+Either a Matchmaking configuration using a Queue
+```ts
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+
+const matchmaking = new gamelift.MatchmakingConfiguration(this, 'Queued Matchmaking', {
+  requestTimeouts: Duration.seconds(35)
+});
+
+const queue = new gamelift.Queue(this, 'GameLift Queue', {
+  placementTimeout: Duration.seconds(10)
+});
+
+matchmaking.withQueue(queue);
+```
+
+The above example implicitly defines the following resources:
+
+* A Queue or a Standalone based Matchmaking configuration
+
+##### RuleSet
+
+Every FlexMatch matchmaker must have a rule set. The rule set determines the two key elements of a match: your game's team structure and size, and how to group players together for the best possible match.
+
+For example, a rule set might describe a match like this: Create a match with two teams of four to eight players each, one team is the cowboy and the other team the aliens. A team can have novice and experienced players, but the average skill of the two teams must be within 10 points of each other. If no match is made after 30 seconds, gradually relax the skill requirements.
+
+```ts
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+
+const ruleSet = new gamelift.MatchmakingRuleSet(this, 'Matchmaking RuleSet');
+
+ruleSet.addPlayerAttribute('skill', PlayerAttributeType.STRING,10
+});
+ruleSet.addTeam('aliens',4,8);
+ruleSet.addTeam('cowboys', 4,8);
+ruleSet.addRule({
+    name: "FairTeamSkill",
+    description: "The average skill of players in each team is within 10 points from the average skill of all players in the match",
+    type: "distance",
+    // get skill values for players in each team and average separately to produce list of two numbers
+    measurements: [ "avg(teams[*].players.attributes[skill])" ],
```

Comment:
I wonder if we can make create a fluent API to help with the construction of these "queries" (not sure if this is the right term). A _very rough_ first approximation:

```ts
const skill = new PlayerAttribute({
  name: 'skill', 
  type: PlayerAttributeType.STRING,
  length: 10,
});

...

measurements: [Query.avg(Teams.all.players[skill])]
```

Users would benefit from auto-complete and it would also catch problems earlier. For example, I suppose it's an error to declare a query like this for "skill" if you haven't defined a player attribute with that name. By making it an object, it forces the user to create it before instead of relying on correlating strings in two (or more) different places.

For a similar mechanism, see the [aws-appsync] module. In particular, the distinction between schema-first and code-first.

[aws-appsync]: https://github.com/aws/aws-cdk/tree/main/packages/%40aws-cdk/aws-appsync